### ISSUE TITLE
Add a native local reranker (TeiRerank) backed by the vendored text-embeddings-inference candle backend

### ISSUE THREAD

**Issue by Jeadie** [spiceai/spiceai#12927]
## Summary

Spice supports remote rerankers only: Cohere, Voyage, Jina, and a generic HTTP reranker (`crates/llms/src/rerank/`). The `Rerank` trait doc comment lists a local BGE reranker as future work. This issue tracks that work.

## Why this is possible now

Spice already links the TEI candle backend as an in-process Rust dependency, not as an external server. `crates/llms/Cargo.toml` pulls in `text-embeddings-backend`, `text-embeddings-backend-candle`, and `text-embeddings-core` from `spiceai/text-embeddings-inference`. `crates/llms/src/embeddings/candle/tei.rs` (`TeiEmbed`) already uses this backend to run local embedding models.

The same backend exposes a classifier path used for cross-encoder reranking:
- `text-embeddings-backend-core`: `ModelType::Classifier`
- `text-embeddings-backend`: `Backend::predict(batch) -> Predictions`

This is the scoring path that reranker models such as `BAAI/bge-reranker-base`, `BAAI/bge-reranker-large`, and `Alibaba-NLP/gte-reranker-modernbert-base` use. TEI's own router and candle backend already have working support and tests for these models.

## Proposed change

Add a `TeiRerank` type in `crates/llms/src/rerank/`, following the load pattern in `TeiEmbed`:
1. Download or read a local reranker model (config, tokenizer, weights) the same way `TeiEmbed` does.
2. Load the model with `ModelType::Classifier` through the vendored candle backend.
3. Implement the `Rerank` trait: call `Backend::predict` for the query/document pairs and return per-document scores.
4. Set `is_remote()` to `false` so callers can use a different concurrency policy than for remote rerankers.
5. Wire it into whatever config path selects a reranker model, alongside Cohere/Voyage/Jina.

## Scope notes

- No new external dependency. No sidecar process. No network call for scoring.
- Follow the existing `Rerank` trait contract: return exactly `documents.len()` scores in input order.

## References

- `crates/llms/src/rerank/mod.rs`
- `crates/llms/src/embeddings/candle/tei.rs`
- `spiceai/text-embeddings-inference` (branch `spiceai`): `backends/core/src/lib.rs`, `backends/src/lib.rs`, `docs/source/en/supported_models.md`

**Comment by Jeadie** [comment]
## Design: native local reranker `TeiRerank` (candle TEI classifier path)

### Goal
Add an in-process reranker backed by the already-vendored `text-embeddings-inference` candle backend, so models like `BAAI/bge-reranker-base`, `BAAI/bge-reranker-large`, and `Alibaba-NLP/gte-reranker-modernbert-base` can score `(query, document)` pairs locally with no sidecar and no network call at scoring time. This is the local-BGE follow-up the `Rerank` trait doc already flags.

### Verified facts from the vendored backend (rev `ac4e457`)
- `text-embeddings-backend-core::ModelType` has exactly two variants at this rev: `Classifier` and `Embedding(Pool)`. Cross-encoder rerankers load as `ModelType::Classifier` (the candle backend threads `model_type` into each architecture's `load`, building a classification head — Bert/Roberta for BGE, ModernBert for gte-reranker).
- `tei_core::infer::Infer::predict(inputs, truncate, truncation_direction, raw_scores, permit) -> ClassificationInferResponse` is the high-level scoring path. `is_classifier()` gates it (`matches!(backend.model_type, ModelType::Classifier)`), so the backend **must** be built with `ModelType::Classifier`.
- `EncodingInput` has a `Dual(String, String)` variant with `From<(String, String)>`. The TEI router's own `/rerank` handler encodes each pair as `(query, text)` and reads `response.results[0]` as the score, with `raw_scores=false` applying **sigmoid** to a single-logit head (or softmax for multi-logit). `TeiRerank` mirrors this exactly for behavioral parity with the tested upstream path.

### Files touched

**1. `crates/llms/src/rerank/tei.rs` (new, `#[cfg(feature = "local_embed")]`)**
- `TeiRerank { infer: Infer, name: String, truncation: Option<TruncationDirection> }`.
- Loaders mirroring `TeiEmbed`: `from_hf(model_id, revision, hf_token, max_seq_length_overwrite, truncation)`, `from_local(...)`, and a shared `from_dir(root, max_seq_length_overwrite, truncation)` that reuses the embeddings candle util helpers (`download_hf_artifacts`, `load_config`, `load_tokenizer`, `position_offset`, `max_seq_length_from_st_config`, `link_files_into_tmp_dir`). The only load difference from `TeiEmbed`: build the backend with `ModelType::Classifier` (no pooling) instead of `ModelType::Embedding(pool)`.
- `impl Rerank`:
  - `rerank(query, docs)`: empty docs → `Ok(vec![])`; otherwise spawn one `infer.predict((query.to_string(), doc.clone()), truncate, dir, raw_scores=false, permit)` per doc via `join_all` (same fan-out shape as `TeiEmbed::embed_futures`), collect `results[0]` in input order. Guards an empty `results` vec with a structured error rather than indexing/panicking.
  - `is_remote() -> false` (the point of the issue — lets UDTF callers pick a local concurrency policy).
  - `model_name()` returns the configured name.
- Load/predict failures map to new `rerank::Error` variants (below); no `unwrap`/`expect`/`panic`.

**2. `crates/llms/src/rerank/mod.rs`**
- `#[cfg(feature = "local_embed")] pub mod tei;` + gated `pub use tei::TeiRerank;`.
- New `Error` variants: `LocalModelLoadFailed { model, source }` and `EmptyPrediction { model }`.
- Update the module doc comment (local BGE reranker is no longer "future").

**3. `crates/llms/src/embeddings/candle/mod.rs`**
- `mod util;` → `pub(crate) mod util;` so the reranker can reuse the loader helpers (already `pub(crate)`) without duplicating them. No behavior change for embeddings.

**4. `crates/spicepod/src/component/rerankers.rs`**
- Add `RerankerPrefix::HuggingFace` and `RerankerPrefix::File`.
- `HuggingFace` reuses `HUGGINGFACE_PATH_REGEX` (same `huggingface:huggingface.co/BAAI/bge-reranker-base[:rev]` convention and revision handling as embeddings/models), so a pinned revision round-trips. `File` strips `file:` to an artifacts directory path.
- Update `TryFrom<&str>`, `Display`, `get_model_id`, the doc comment, and add unit tests (prefix recognition, model-id extraction incl. revision, HF regex round-trip).

**5. `crates/runtime/src/model/rerank.rs`**
- Add `HuggingFace` / `File` match arms, gated `#[cfg(feature = "models")]` (the runtime `models` feature enables `llms/local_embed`). HuggingFace splits the id with `split_hf_model_id` and calls `TeiRerank::from_hf`; File calls `TeiRerank::from_dir`. `#[cfg(not(feature = "models"))]` arms return the existing `UnknownSource`-style error so a non-`models` build degrades gracefully.
- New loader-error variants surfaced through the existing `BuildFailedSnafu` shape.

### Trade-offs / assumptions (to confirm in review — no local `cargo` in this sandbox)
- **`results[0]` + sigmoid**: chosen for exact parity with TEI's own `/rerank` router, which is the tested path for these models. A multi-logit classifier would yield a softmaxed class-0 probability; documented, and BGE/gte rerankers are single-logit so this is correct for the models the issue names.
- **Per-pair `predict` calls via `join_all`**: matches both `TeiEmbed` and the upstream router; the `Infer` queue batches internally, so this is not N sequential round-trips.
- **`local_embed` reuse**: the reranker rides the existing embeddings feature/dep set rather than adding a new one, per "no new external dependency."
- I could not compile-check locally; types/signatures were read directly from the vendored source at the pinned rev and existing call sites. CI/review to catch any residual mismatch.

### Test plan
- Unit: spicepod prefix/model-id parsing (incl. HF revision round-trip and delimiter rejection); `rerank::Error` mapping.
- Integration (follow-up, needs model download): a `local`/`-federated` reranker spicepod loading `BAAI/bge-reranker-base` via `huggingface:`, asserting `documents.len()` scores in input order and that a clearly-relevant doc outscores an irrelevant one.

**Comment by Jeadie** [comment]
## Implemented

Code-complete per the design above. Files changed:

- **`crates/llms/src/rerank/tei.rs`** (new) — `TeiRerank`: `from_hf` / `from_local` / `from_dir` load the model with `ModelType::Classifier` through the vendored candle backend, reusing the embeddings candle loader helpers. `rerank()` fans out one `Infer::predict((query, doc), …, raw_scores=false, …)` per document (each future owns a cheap `Infer` clone, like `TeiEmbed`), reads `results[0]` (sigmoid-calibrated), and returns `documents.len()` scores in input order. `is_remote() → false`.
- **`crates/llms/src/rerank/mod.rs`** — `#[cfg(feature = "local_embed")]`-gated `pub mod tei` + `pub use TeiRerank`; new `Error::LocalModelLoadFailed` / `Error::EmptyPrediction`; refreshed the trait doc (local BGE is no longer "future").
- **`crates/llms/src/embeddings/candle/mod.rs`** — `mod util` → `pub(crate) mod util` so the reranker reuses the loader helpers instead of duplicating them (no behavior change).
- **`crates/spicepod/src/component/rerankers.rs`** — added `RerankerPrefix::HuggingFace` (reuses `HUGGINGFACE_PATH_REGEX`, revision round-trips) and `RerankerPrefix::File`; updated `TryFrom`/`Display`/`get_model_id`/doc + unit tests.
- **`crates/runtime/src/model/rerank.rs`** — `#[cfg(feature = "models")]` arms build `TeiRerank` for `huggingface:`/`file:` (optional `max_seq_length`, `truncate: none|end|start` params); `#[cfg(not(feature = "models"))]` returns a clear "rebuild with `models`" error.

Usage:

```yaml
rerankers:
  - from: huggingface:huggingface.co/BAAI/bge-reranker-base
    name: local_reranker
    params:
      truncate: end        # optional: none (default) | end | start
      # max_seq_length: 512 # optional
```

Assumptions (unchanged from the design; no local `cargo` in this env, so CI/review is the compile gate): `results[0]` + sigmoid mirrors TEI's own `/rerank` router (correct for single-logit BGE/gte cross-encoders); per-pair `predict` relies on `Infer`'s internal batching; the reranker rides the existing `local_embed`/`models` feature set — no new dependency.
